rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/metadata/CalculateDeltaNode.java

```diff
@@ -74,51 +76,53 @@ public void init(TbContext ctx, TbNodeConfiguration configuration) throws TbNode
 
     @Override
     public void onMsg(TbContext ctx, TbMsg msg) {
-        if (msg.getType().equals(SessionMsgType.POST_TELEMETRY_REQUEST.name())) {
-            JsonNode json = JacksonUtil.toJsonNode(msg.getData());
-            String inputKey = config.getInputValueKey();
-            if (json.has(inputKey)) {
-                DonAsynchron.withCallback(getLastValue(msg.getOriginator()),
-                        previousData -> {
-                            double currentValue = json.get(inputKey).asDouble();
-                            long currentTs = msg.getMetaDataTs();
-
-                            if (useCache) {
-                                cache.put(msg.getOriginator(), new ValueWithTs(currentTs, currentValue));
-                            }
-
-                            BigDecimal delta = BigDecimal.valueOf(previousData != null ? currentValue - previousData.value : 0.0);
-
-                            if (config.isTellFailureIfDeltaIsNegative() && delta.doubleValue() < 0) {
-                                ctx.tellNext(msg, TbRelationTypes.FAILURE);
-                                return;
-                            }
-
-
-                            if (config.getRound() != null) {
-                                delta = delta.setScale(config.getRound(), RoundingMode.HALF_UP);
-                            }
-
-                            ObjectNode result = (ObjectNode) json;
-                            if (delta.stripTrailingZeros().scale() > 0) {
-                                result.put(config.getOutputValueKey(), delta.doubleValue());
-                            } else {
-                                result.put(config.getOutputValueKey(), delta.longValueExact());
-                            }
-
-                            if (config.isAddPeriodBetweenMsgs()) {
-                                long period = previousData != null ? currentTs - previousData.ts : 0;
-                                result.put(config.getPeriodValueKey(), period);
-                            }
-                            ctx.tellSuccess(TbMsg.transformMsg(msg, msg.getType(), msg.getOriginator(), msg.getMetaData(), JacksonUtil.toString(result)));
-                        },
-                        t -> ctx.tellFailure(msg, t), ctx.getDbCallbackExecutor());
-            } else {
-                ctx.tellNext(msg, "Other");
-            }
-        } else {
+        if (!msg.getType().equals(SessionMsgType.POST_TELEMETRY_REQUEST.name())) {
             ctx.tellNext(msg, "Other");
+            return;
         }
+        JsonNode json = JacksonUtil.toJsonNode(msg.getData());
+        if (!json.isObject()) {
+            throw new IllegalArgumentException("Message body is not an object!");
+        }
+        String inputKey = config.getInputValueKey();
+        if (!json.has(inputKey)) {
+            ctx.tellNext(msg, "Other");
+            return;
+        }
+        withCallback(getLastValue(msg.getOriginator()),
```

Comment:
The cache is enabled by default. Also, we use PostgreSQL to store the latest values in 98% of environments. I suggest making the following improvement:
1) If we use a cache, we don't make async calls to the DB to ensure the correct order of messages (even without the SeqByOriginator). We will need to add a blocking call to the *org.thingsboard.server.dao.timeseries.TimeseriesService*
2) If we don't use the cache, we may keep async logic